Assets/Scripts/Combat/Abilities/MagicShield.cs

```diff
@@ -0,0 +1,269 @@
+    private Timer SequenceTimer;
+    private Queue<Button> Sequence = new Queue<Button>();
+    private Queue<Button> SequenceOrder = new Queue<Button>();
+    private Queue<GameObject> Directions = new Queue<GameObject>();
+
+    [SerializeField]
+    private GameObject Shield;
+    [SerializeField]
+    private GameObject[] Arrows; // 0: Up, 1: Down, 2: Left, 3: Right 
+
+    private readonly float SequenceDuration = 2f;
+    private readonly float InputDuration = 5f;
+    private readonly float PositionOffset = 4f;
+    private Vector3 Position = Vector3.zero;
+    private Phase CurrentPhase = Phase.Inactive;
+
+    private int CorrectInputs = 0;
+    private readonly int MagicShieldFactor = 20;
+    private int MagicShieldHealth;
+
+    public new void Start()
+    {
+        base.Start();
+        SequenceTimer = GetComponent<Timer>();
+        Shield.SetActive(false);
+
+        foreach(GameObject arrow in Arrows)
+            arrow.SetActive(false);
+
+        TargetSchema = new TargetSchema(
+            0,
+            CombatantType.Ally,
+            SelectorType.All);
+    }
+
+    new public void StartAbility()
+    {
+        CorrectInputs = 0;
+        base.StartAbility();
+    }
+
+    private void Update()
+    {
+        if(CurrentPhase == Phase.SequencePhase)
+            SequencePhaseUpdate();
+
+        else if(CurrentPhase == Phase.InputSequence)
+            InputSequenceUpdate();
+    }
+
+    protected override void ContinueAbilityAfterTargeting()
+    {
+        StartMagicShieldSequence();
+    }
+
+    protected override void EndAbility()
+    {
+        Debug.Log($"Magic Shield Total Health: {MagicShieldHealth}");
+
+        for (int i = 0; i < TargetedCombatants.Length; i++)
+            TargetedCombatants[i].GetComponent<Combatant>().ApplyShield(MagicShieldHealth);
+
+        CombatSystem.EndTurn(this.GetComponentInParent<Combatant>().gameObject);
+
+    }
+
+    private void StartMagicShieldSequence()
+    {
+        Debug.Log("Starting Magic Shield Ability");
+
+        Shield.transform.position = GameObject.FindGameObjectWithTag("Player").transform.position;
+        Shield.transform.position -= new Vector3(0, 5, 0);
+        Shield.transform.Translate(new Vector3(0, 10, 0));
+
+        Array buttons = Enum.GetValues(typeof(Button));
+
+        for (int i = 0; i < buttons.Length; i++)
+        {
+            Button random_button = (Button)buttons.GetValue(Random.Range(0, buttons.Length));
+            Sequence.Enqueue(random_button);
+            SequenceOrder.Enqueue(random_button);
+        }
+
+        Position = Arrows[0].transform.position;
+        Debug.Log("Sequence Appearing!");
+        SequenceTimer.StartTimer(SequenceDuration);
+        CurrentPhase = Phase.SequencePhase;
+          
+    }
+
+    private void SequencePhaseUpdate()
+    {
+
+        if (SequenceTimer.IsInProgress())
+        {
+            if (Sequence.Count != 0)
+            {
```

Comment:
You could technically merge those into a single if

Reply:
Fixed!